### ISSUE TITLE
Write a plugin in vim9script to be compatible with the original Vim

### ISSUE THREAD

**Issue by jclsn** [apple/pkl-neovim#11]
This plugin exclusively written for Neovim, which can't be used in the original Vim. Many people still use it though. Would be great if there could also get one written in vim9script for Vim!

**Comment by holzensp** [comment]
That's a totally reasonable ask. It's going to be a while before we can prioritize this, though, so if you want it fast; PRs are most welcome ;) Ideally, though, we have one plugin that works in either, if that's doable.

**Comment by jclsn** [comment]
@holzensp Thank you for being open for this. It is doable, but currently it would have to be written in legacy vimscript. Although I am not sure if Neovim will add vim9script support in the future.

Ah seems there are plans

https://neovim.discourse.group/t/will-vim9-be-supported/2229/10

**Comment by jclsn** [comment]
Upon further inquiry it seems they are not likely to support it any time soon.

https://github.com/neovim/neovim/issues/13625

You could still write the plugin in legacy vimscript though, to make it available for both editors.

**Comment by bioball** [comment]
That makes sense! I'd be fine with structuring this repo so that it's still usable for only vim users (where features like tree-sitter highlighting and LSP are disabled).

> In fact I already started doing that [here](https://github.com/jclsn/pkl-vim/blob/master/syntax/pkl.vim). Feel free to fork and improve things. I don't see myself going all the way here.

Very cool! Happy to accept PRs for this.

**Comment by jclsn** [comment]
So I made some progress. What are these annotations/decorators like @SourceCode? I could not find them in the documentation. In VSCode they don't seem to get highlighted either. I also highlighted the # comments inside verbatim strings as comments. This isn't done in VSCode either. Also nothing in the docs about those.

**Comment by jclsn** [comment]
I am just matching the @whatever now and highlighting it. Looks good so far.

I interpreted the # comments as a way to comment inside raw strings. There are some patterns in those raw strings and the text after the # can't be part of the pattern. It's a bit confusing.

Good that you mention the shebang. Going to add it!

**Comment by jclsn** [comment]
Yes, but inside them are these hash comments, which don't seem to be part of the pattern.

https://github.com/apple/pkl.tmbundle/blob/af454a169c7d328d4838e7f33fff111eb8e33967/pkl/pkl.tmLanguage.pkl#L42

**Comment by bioball** [comment]
Ah, those are regex comments. They're not actually part of the pattern, they're just comments to help maintainers.

**Comment by jclsn** [comment]
Okay, so those regex comments are not part of pkl and therefore should not be included in the highlighting, I reckon?

**Comment by bioball** [comment]
Our grammar for strings are borrowed from [Swift](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/stringsandcharacters/#Extended-String-Delimiters), which works exactly the same way.

Two pounds and up are very rare, but they make sense when you want to write strings that otherwise need escaping. For example, if your string contains verbatim `\#`, you might want to use two pounds.

**Comment by jclsn** [comment]
Doesn't seem to be supported for Swift in the official Vim repo: 

https://github.com/vim/vim/blob/41ee98c3c5342867cb99dfcddbe8d53caeda22db/runtime/syntax/swift.vim#L183

I guess I could just go with a maximum of four pounds. Do you support highlighting for an infinite amount of pounds in treesitter?
